File: 백준/Gold/7576. 토마토/토마토.py
from collections import deque

M, N = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(N)]
visited = [[0]*M for _ in range(N)]
dx, dy = [0, 1, 0, -1], [1, 0, -1, 0]
# print(arr)

def bfs(lst):
    q = deque(lst)
    while q:
        x, y = q.popleft()
        for i in range(4):
            nx, ny = x+dx[i], y+dy[i]
            if 0<=nx<N and 0<=ny<M and visited[nx][ny] == 0 and arr[nx][ny] != -1:
                q.append((nx, ny))
                visited[nx][ny] = 1
                arr[nx][ny] = arr[x][y] + 1
    mx = 0
    for i in range(N):
        for j in range(M):
            if arr[i][j] == 0:
                return -1
            mx = max(mx, arr[i][j]-1)
    return mx


even = []
for i in range(N):
    for j in range(M):
        if arr[i][j] == 1:
            even.append((i, j))
            visited[i][j] = 1

print(bfs(even))

File: 백준/Gold/7576. 토마토/토마토.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int M, N;

int dx[4] = {0, 1, 0, -1};
int dy[4] = {1, 0, -1, 0,};

vector<vector<int>> arr;
vector<vector<int>> visited;
queue<pair<int, int>> q;

int bfs(){
    int total = 0;
    while(!q.empty()){
        auto[x, y] = q.front();
        q.pop();
        total = max(total, visited[x][y]);

        for(int i=0; i<4; i++){
            int nx = x + dx[i];
            int ny = y + dy[i];
            if(0<=nx && nx<N && 0<=ny && ny<M && arr[nx][ny]==0 && visited[nx][ny]==0){
                q.push({nx, ny});
                visited[nx][ny] = visited[x][y] + 1;
                arr[nx][ny] = 1;
            }
        }
    }
    for(int i=0; i<N; i++){
        for(int j=0; j<M; j++){
            if(arr[i][j] == 0){
                return -1;
            }
        }
    }

    return total-1;
}


int main(void){
    cin >> M >> N;

    arr = vector<vector<int>>(N, vector<int>(M, 0));
    visited = vector<vector<int>>(N, vector<int>(M, 0));

    for(int i=0; i<N; i++){
        for(int j=0; j<M; j++){
            cin >> arr[i][j];
            if(arr[i][j]==1){
                q.push({i, j});
                visited[i][j] = 1;
            }
        }
    }

    cout << bfs() << endl;

    return 0;
}